chore(*): Improve typing of Tokens (#963)
johnsonandjohnson/Bodiless-JS: packages/fclasses/doc/Types.md

# Tokens and Higher Order Components in Typescript

The Bodiless design system expresses tokens as React higher order components,
which can be a bit tricky to type correctly. Fortunately, Bodiless exposes
a number of utility types which simplify the process.

Before proceding, you should have a solid understanding of
[higher order compoennts in React](https://reactjs.org/docs/higher-order-components.html)
as well as the basics of how to type them traditionally.
[This excellent article by James Ravenscroft](https://medium.com/@jrwebdev/react-higher-order-component-patterns-in-typescript-42278f7590fb)
is a good introduction to the latter. You should also have read and be familiar
with the BodilessJS documentation on
[the FClasses Design API](../Architecture/FClasses).

## A few terms to begin with:

- **Higher Order Component** A function which takes one component and returns
  a new one.
- **Token** A unit of design or behavior, expressed as a higher order component.
- **Base Component** The component to which a token or HOC applies.
- **Enhanced Component** The component returned by a token or HOC.

> - **Enhancer** A token or HOC which introducres new props in the enhanced

Author: ```suggestion
- **Enhancer** A token or HOC which introduces new props in the enhanced
```

>   compoennt.

Author: ```suggestion
  component.
```

- **Injector** A token or HOC which provides a value for one of the base
  component's props. That prop becomes optional in the enhanced component.
- **Token Generator** A function which takes some parameters and returns a token
  or HOC.

> Note the terms 'injector' and 'enhancer' are borrowed from
> [this excellent article](https://medium.com/@jrwebdev/react-higher-order-component-patterns-in-typescript-42278f7590fb)
> by James Ravenscroft.

## Generic Tokens

The simplest type of token is one which does not alter the props of the base
component.  For this we use the `Token` type. For example, if we have the
following base types:
```ts
type BaseProps = {
  foo: string,
  bar: string,
};
const Base: FC<BaseProps> = () => null;

type Base2Props = {
  baz: string,
};
const Base2: FC<Base2Props> = () => null;

const Base3: FC<BaseProps & Base2Props> = () => null;
```

> Then we might define a generic token generator producing a token which iinserts

Author: ```suggestion
Then we might define a generic token generator producing a token which inserts
```

a child as:
```ts
const withChild = (Child: ComponentOrTag<any>): Token => C => {
  const WithChild: FC<any> = props => <C {...props}><Child /></C>;
  return WithChild;
};
```
This will ensure that the enhanced component has the same signature
as the base component:
```ts
const A1 = withChild('span')(Base);
const A1T1 = <A1 foo="foo" bar="bar" />;
// @ts-expect-error missing prop bar
const A1T2 = <A1 foo="foo" />;
// @ts-expect-error prob baz is undefined
const A1T3 = <A1 foo="foo" bar="bar" baz="baz" />;
```
Some things to note:

> - We give the enhanced componennt an explicit name (`Withhild`) so that it is

Author: ```suggestion
- We give the enhanced component an explicit name (`Withld`) so that it is
```
